ata/src/config.rs

```diff
@@ -0,0 +1,250 @@
+
+lazy_static! {
+    static ref DEFAULT_CONFIG_FILENAME: PathBuf = "ata.toml".into();
+}
+
+// For definitions, see https://platform.openai.com/docs/api-reference/completions/create
+#[repr(C)]
+#[derive(Clone, Deserialize, Debug, Serialize, Reflect)]
+#[serde(default)]
+pub struct Config {
+    pub api_key: String,
+    pub model: String,
+    pub max_tokens: i64,
+    pub temperature: f64,
+    pub suffix: Option<String>,
+    pub top_p: f64,
+    pub n: u64,
+    pub stream: bool,
+    pub logprobs: u8,
+    pub echo: bool,
+    pub stop: Vec<String>,
+    pub presence_penalty: f64,
+    pub frequency_penalty: f64,
+    pub best_of: u64,
+    pub logit_bias: HashMap<String, f64>,
+}
+
+impl Config {
+    pub fn validate(&self) -> Result<(), String> {
+        if self.api_key.is_empty() {
+            return Err(String::from("API key is missing"));
+        }
+
+        if self.model.is_empty() {
+            return Err(String::from("Model ID is missing"));
+        }
+
+        if self.max_tokens < 1 || self.max_tokens > 2048 {
+            return Err(String::from("Max tokens must be between 1 and 2048"));
+        }
+
+        if self.temperature < 0.0 || self.temperature > 1.0 {
+            return Err(String::from("Temperature must be between 0.0 and 1.0"));
+        }
+
+        if let Some(suffix) = &self.suffix {
+            if suffix.is_empty() {
+                return Err(String::from("Suffix cannot be an empty string"));
+            }
+        }
+
+        if self.top_p < 0.0 || self.top_p > 1.0 {
+            return Err(String::from("Top-p must be between 0.0 and 1.0"));
+        }
+
+        if self.n < 1 || self.n > 10 {
+            return Err(String::from("n must be between 1 and 10"));
+        }
+
+        if self.logprobs > 2 {
+            return Err(String::from("logprobs must be 0, 1, or 2"));
+        }
+
+        if self.stop.iter().any(|stop| stop.is_empty()) || self.stop.len() > 4 {
+            return Err(String::from("Stop phrases cannot contain empties"));
+        }
+
+        if self.presence_penalty < 0.0 || self.presence_penalty > 1.0 {
+            return Err(String::from("Presence penalty must be between 0.0 and 1.0"));
+        }
+
+        if self.frequency_penalty < 0.0 || self.frequency_penalty > 1.0 {
+            return Err(String::from(
+                "Frequency penalty must be between 0.0 and 1.0",
+            ));
+        }
+
+        if self.best_of < 1 || self.best_of > 5 {
+            return Err(String::from("best_of must be between 1 and 5"));
+        }
+
+        for (key, value) in &self.logit_bias {
+            if value < &-2.0 || value > &2.0 {
+                return Err(format!(
+                    "logit_bias for {} must be between -2.0 and 2.0",
+                    key
+                ));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            model: "text-davinci-003".into(),
```

Comment:
Things move quickly in OpenAI API's and since the release of the ChatGPT models, I think those should be the primary. I think that v2 of `ata` should just drop support for non-chat models for simplicity. What do you think?

Reply:
I also wonder whether people actually would use the settings such as `top_p`. For people who want to use that, there is https://platform.openai.com/playground. `ata` aims more at productivity.

Reply:
Well that's why I added the ability to also specify configuration files by name. That way you can load an alternate one for a certain project.

Reply:
Actually I wanted to make it so that every single configuration argument could just as well be specified via the command line or the operating system environment and then in that case you don't even need to use configuration files. So the order of loading would be, configuration files if any, environment, arguments.

Reply:
> I think that v2 of ata should just drop support for non-chat models for simplicity. What do you think?

Sorry that's really basically the opposite of what I had in mind. So the reason that this project was done on Android (part of the idea was that I would only actually work on things that would make using a rooted Android phone better) is that the playground website is extremely broken in the Android browsers. The streaming feature barely works, it always loses your text and doesn't get it back through undo etc.

I also wanted to be able to pipe to ata. I did not quite finish that yet but I will.

As I see it, ata is most useful as a _generic_ console application which supports all of the GPT APIs. And much like most other console applications, that means that it has an interactive mode and a batch mode.

I think that if you want to use ChatGPT, even by default, that's no problem. But I would not drop support for the existing API, I would just put it behind an option.

Reply:
Thank you for giving more details. I have been thinking about it for a bit more and I see your point. For me, the main goal of `ata` was also to provide a more productive interface to ChatGPT than is currently available at chat.openai.com.

> As I see it, ata is most useful as a _generic_ console application which supports all of the GPT APIs. And much like most other console applications, that means that it has an interactive mode and a batch mode.
> 
> I think that if you want to use ChatGPT, even by default, that's no problem. But I would not drop support for the existing API, I would just put it behind an option.

Yes I agree that that would be best. The problem for me is that I do not have time to support all GPT APIs. In the ideal case, I would start adding tests for example. I currently do not have time for that since my PhD project also has a lot of deadlines currently. At the same time, I do want to try to keep up with OpenAI while they are moving at lightning speed. So, the most useful middle ground there in my opinion is then to drop support for older models and only support the, much cheaper, chatgpt models. That way, Android and terminal users can have productive interactions with at least one GPT model

Reply:
> In the ideal case, I would start adding tests for example.

More specifically, the complexity of the project is nearing my own capabilities. Without tests, I cannot increase complexity much more or I would start introducing more and more bugs. However, tests are complex to make too since it would require a mock stdout and a mock OpenAI server. Both are possible, but take time which I don't have

Reply:
That's fine, I think we should just fork. I'm not really going to add test because I don't really care if it works or not for every user in every use case... They can report issues and then I'll fix those. My plan certainly is not to have a mock server and a mock standard out.